mailserver/mailserver.go

```diff
@@ -531,15 +544,28 @@ func (s *WMailServer) processRequestInBundles(
 
 	// Publish
 	for _, batch := range batches {
-		output <- batch
+		select {
+		case output <- batch:
+		// It might happen that during producing the batches,
+		// the connection with the peer goes down and
+		// the consumer of `output` channel exits prematurely.
+		// In such a case, we should stop pushing batches and exit.
+		case <-cancel:
+			log.Error("[mailserver:processRequestInBundles] failed to push all batches",
+				"requestID", requestID)
+			break
+		case <-time.After(timeout):
```

Comment:
i noticed discussion about this... but it feels really weird. there is a finite conditions that needs to be taken into account. if consumer stopped this loop needs to stop, otherwise it should work. literally no reason to add such hack

Reply:
No reason if this or underlying code will never change and never get any regressions in the future.

Reply:
like why not to add timeout to every select block :) ?

Reply:
I actually think that using lots of selects and channels with no timeouts is a bad coding practice, especially if your code supposed to be run for days and months with no restarts.

Reply:
oh, really? i think you need to define concrete conditions when select statement will terminate. and it doesn't have to be limited by a timer. for instance if you look at p2p/server.go, most of the select statements never can be terminated by a timer, it would be just wrong

Reply:
i can imagine different pattern, where whole DeliverMail procedure is time-limited, but it would be implemented differently

Reply:
I agree that this case should never happen and `case <-cancel` is here to cover stopped peers. However, we don't own devp2p stack and apparently we don't trust it. We can run this code for a few days and if this log never appears (which I believe is the case), this fragment can be removed.

> I actually think that using lots of selects and channels with no timeouts is a bad coding practice, especially if your code supposed to be run for days and months with no restarts.

If the code is well architected, I think timeouts are not needed in such cases. They may appear at some entry points handling incoming requests or explicitly timing out network connections if there is no way to timeout the connection itself due to some reason.

Reply:
Sure. But if you have an assumption that some code should complete in max N minutes after some event, having a watchdog for it both helps to identifying issues and mitigating consequences.
For instance, no modern mobile OS will let you block the rendering loop for too long. Same is true for most modern OSes. It wasn't true for Windows 95 though, and we all seen the consequences.

Reply:
> If the code is well architected, I think timeouts are not needed in such cases.

Yeah, if there is no potential for a deadlock then it is obviously not needed. Unfortunately, in Go, when using channels you are asking for trouble :-P

Reply:
> I agree that this case should never happen and `case <-cancel` is here to cover stopped peers. However, we don't own devp2p stack and apparently we don't trust it.

the bug wasn't in devp2p code though :)

> We can run this code for a few days and if this log never appears (which I believe is the case), this fragment can be removed.

yeah i think that it won't appear too. unless 1 minute is not enough to deliver data for some of the congested mail servers, like it was with HK mail servers. but if that happens, isn't this timeout does is harmful?

Reply:
> the bug wasn't in devp2p code though :)

True but we still don't have a clear picture how devp2p behaves in the case of half-open TCP connections or network switches.

> but if that happens, isn't this timeout is harmful?

I don't see anything harmful here. It will close `bundles` channel which will exit goroutine processing it and `DeliverMail` will exit normally. The only bad thing that happens is that not all bundles (more precisely older envelopes) will get delivered.

Reply:
> True but we still don't have a clear picture how devp2p behaves in the case of half-open TCP connections or network switches.

i see that either side will be eventually closed by read/write deadline. it seems pretty straightforward. what do you think is confusing?

Reply:
It seems :) here there is no explicit timeout when sending envelopes, so that part is confusing -- after what time it times out.

Reply:
it will time out after deadline passed since actual write started

Reply:
It's still implicit. Can you give me precise duration here?

There are a few logs so this condition is possible: https://kibana.status.im/goto/4b14a7de6e58a9237a18657beb94712b. On the other hand, these logs are old and comes from a period of time when we were deploying stuff.

Reply:
it is hard to give precise duration. all writes are processed sequentially, and the deadline is counted from the time message hits socket. if there are no other writes - single write can take at most 20s.

> There are a few logs so this condition is possible: https://kibana.status.im/goto/4b14a7de6e58a9237a18657beb94712b. On the other hand, these logs are old and comes from a period of time when we were deploying stuff.

got it. but it is still possible on a slow connection. e.g. if there are 10 batches, and every single batch takes 19s to deliver.. in the end it will take longer than a minute